File: crates/ward_runtime/py/wardscript/providers/anthropic.py
"""Claude, through the Anthropic API. The return type's schema is given as a tool the
model must call, so answers are structured; usage is reported to budgets."""

from __future__ import annotations

import json
from typing import Any, Iterator

from ..model import AiRequest, Completion
from . import cost, object_schema, prompt_text, provider_errors

DEFAULT_MODEL = "claude-sonnet-5"


class Anthropic:
    #: `stream` yields the tool input, `{"value": ...}`.
    stream_wraps_value = True

    def __init__(
        self,
        model: str | None = None,
        *,
        max_tokens: int = 4096,
        prices: tuple[float, float] | None = None,
        client: Any = None,
    ) -> None:
        """`prices` are dollars per million input and output tokens, for `cost` budgets;
        without them the cost is unknown, and a `cost` budget refuses the call. `client` defaults to `anthropic.Anthropic()`,
        which reads `ANTHROPIC_API_KEY`, without the SDK's own retries: the runtime
        retries by the `ai fn`'s model policy."""
        if client is None:
            import anthropic

            client = anthropic.Anthropic(max_retries=0)
        self.client = client
        self.model = model or DEFAULT_MODEL
        self.max_tokens = max_tokens
        self.prices = prices

    def _request(self, request: AiRequest) -> dict[str, Any]:
        return {
            "model": self.model,
            "max_tokens": self.max_tokens,
            "messages": [{"role": "user", "content": prompt_text(request)}],
            "tools": [
                {
                    "name": "answer",
                    "description": f"Give the answer of `{request.function}`.",
                    "input_schema": object_schema(request.schema),
                }
            ],
            "tool_choice": {"type": "tool", "name": "answer"},
        }

    def _completion(self, tool_input: Any, text: str, input_tokens: int, output_tokens: int) -> Completion:
        if isinstance(tool_input, dict) and "value" in tool_input:
            text = json.dumps(tool_input["value"], ensure_ascii=False)
        # Else no tool call: whatever text there is gets decoded, and rejected if wrong.
        return Completion(
            text,
            tokens=input_tokens + output_tokens,
            cost=cost(self.prices, input_tokens, output_tokens),
        )

    def complete(self, request: AiRequest) -> Completion:
        with provider_errors():
            message = self.client.messages.create(**self._request(request))
        answer = next((b for b in message.content if getattr(b, "type", None) == "tool_use"), None)
        text = "".join(getattr(b, "text", "") for b in message.content)
        usage = message.usage
        return self._completion(
            answer.input if answer is not None else None, text, usage.input_tokens, usage.output_tokens
        )

    def stream(self, request: AiRequest) -> Iterator[str | Completion]:
        """Yields the tool call's JSON (`{"value": ...}`) as it arrives, then the answer."""
        with provider_errors():
            yield from self._stream(request)

    def _stream(self, request: AiRequest) -> Iterator[str | Completion]:
        events = self.client.messages.create(**self._request(request), stream=True)
        tool_json, text, input_tokens, output_tokens = "", "", 0, 0
        for event in events:
            kind = getattr(event, "type", None)
            if kind == "message_start":
                input_tokens = event.message.usage.input_tokens
                output_tokens = getattr(event.message.usage, "output_tokens", 0) or 0
            elif kind == "message_delta":
                output_tokens = event.usage.output_tokens
            elif kind == "content_block_delta":
                delta = event.delta
                if delta.type == "input_json_delta":
                    tool_json += delta.partial_json
                    yield delta.partial_json
                elif delta.type == "text_delta":
                    text += delta.text
                    yield delta.text
        try:
            tool_input = json.loads(tool_json) if tool_json else None
        except json.JSONDecodeError:
            tool_input = None
        yield self._completion(tool_input, text or tool_json, input_tokens, output_tokens)

File: crates/ward_runtime/ts/src/providers/anthropic.ts
// Claude, through the Anthropic Messages API (with `fetch`, no SDK). The return
// type's schema is a tool the model must call, so answers are structured.

import type { AiRequest, Completion } from "../model.ts";
import { cost, objectSchema, post, promptText } from "./common.ts";

export const DEFAULT_MODEL = "claude-sonnet-5";

export interface AnthropicOptions {
  maxTokens?: number;
  /** Dollars per million input and output tokens, for `cost` budgets. */
  prices?: [number, number] | null;
  /** Defaults to `ANTHROPIC_API_KEY`. */
  apiKey?: string;
  baseUrl?: string;
}

export class Anthropic {
  readonly model: string;
  readonly prices: [number, number] | null;
  private readonly options: AnthropicOptions;

  constructor(model: string = DEFAULT_MODEL, options: AnthropicOptions = {}) {
    this.model = model;
    this.prices = options.prices ?? null;
    this.options = options;
  }

  async complete(request: AiRequest): Promise<Completion> {
    const key = this.options.apiKey ?? process.env.ANTHROPIC_API_KEY ?? "";
    const data = await post(
      `${this.options.baseUrl ?? "https://api.anthropic.com"}/v1/messages`,
      { "x-api-key": key, "anthropic-version": "2023-06-01" },
      {
        model: this.model,
        max_tokens: this.options.maxTokens ?? 4096,
        messages: [{ role: "user", content: promptText(request) }],
        tools: [{ name: "answer", description: `Give the answer of \`${request.function}\`.`, input_schema: objectSchema(request.schema) }],
        tool_choice: { type: "tool", name: "answer" },
      },
    );
    const content = (data.content ?? []) as { type: string; input?: { value?: unknown }; text?: string }[];
    const call = content.find((b) => b.type === "tool_use");
    const usage = (data.usage ?? {}) as { input_tokens?: number; output_tokens?: number };
    const input = usage.input_tokens ?? 0;
    const output = usage.output_tokens ?? 0;
    const text =
      call?.input && "value" in call.input
        ? JSON.stringify(call.input.value)
        : content.map((b) => b.text ?? "").join("");
    return { text, tokens: input + output, cost: cost(this.prices, input, output) };
  }
}
